logstreams/src/test/java/io/zeebe/logstreams/state/ReplicateSnapshotControllerTest.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.logstreams.state;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.groups.Tuple.tuple;
+
+import io.zeebe.db.impl.DefaultColumnFamily;
+import io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory;
+import io.zeebe.logstreams.processor.SnapshotChunk;
+import io.zeebe.logstreams.processor.SnapshotReplication;
+import io.zeebe.logstreams.util.RocksDBWrapper;
+import io.zeebe.test.util.AutoCloseableRule;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class ReplicateSnapshotControllerTest {
```

Comment:
Please add more test cases to cover the not so happy paths:
* fail to replicate snapshot file
* no snapshot to replicate
* receive snapshot and directory exists already
* fail to write snapshot file
* receive not all files of a snapshot

Reply:
* fail to write snapshot file  -> then snapshot will not become valid (follower needs to wait for next replicated snapshot)
* receive not all files of a snapshot ->  then snapshot will not become valid

Testing that something does not happen is hard (e.g.  "snapshot will not become valid")
For the other cases I tried to add appropriate tests.